src/Interpreters/JIT/CHJIT.cpp

```diff
@@ -0,0 +1,388 @@
+#include "CHJIT.h"
+
+#if USE_EMBEDDED_COMPILER
+
+#include <llvm/Analysis/TargetTransformInfo.h>
+#include <llvm/IR/BasicBlock.h>
+#include <llvm/IR/DataLayout.h>
+#include <llvm/IR/DerivedTypes.h>
+#include <llvm/IR/Function.h>
+#include <llvm/IR/IRBuilder.h>
+#include <llvm/IR/Mangler.h>
+#include <llvm/IR/Type.h>
+#include <llvm/IR/LegacyPassManager.h>
+#include <llvm/ExecutionEngine/JITSymbol.h>
+#include <llvm/ExecutionEngine/SectionMemoryManager.h>
+#include <llvm/ExecutionEngine/JITEventListener.h>
+#include <llvm/MC/SubtargetFeature.h>
+#include <llvm/Support/DynamicLibrary.h>
+#include <llvm/Support/Host.h>
+#include <llvm/Support/TargetRegistry.h>
+#include <llvm/Support/TargetSelect.h>
+#include <llvm/Transforms/IPO/PassManagerBuilder.h>
+#include <llvm/Support/SmallVectorMemoryBuffer.h>
+
+#include <Common/Exception.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int CANNOT_COMPILE_CODE;
+    extern const int LOGICAL_ERROR;
+}
+
+/** Simple module to object file compiler.
+  * Result object cannot be used as machine code directly, it should be passed to linker.
+  */
+class JITCompiler
+{
+public:
+
+    explicit JITCompiler(llvm::TargetMachine &target_machine_)
+    : target_machine(target_machine_)
+    {
+    }
+
+    std::unique_ptr<llvm::MemoryBuffer> compile(llvm::Module & module)
+    {
+        auto materialize_error = module.materializeAll();
+        if (materialize_error)
+        {
+            std::string error_message;
+            handleAllErrors(std::move(materialize_error),
+                [&](const llvm::ErrorInfoBase & error_info) { error_message = error_info.message(); });
+
+            throw Exception(ErrorCodes::CANNOT_COMPILE_CODE, "Cannot materialize module {}", error_message);
+        }
+
+        llvm::SmallVector<char, 4096> object_buffer;
+
+        llvm::raw_svector_ostream object_stream(object_buffer);
+        llvm::legacy::PassManager pass_manager;
+        llvm::MCContext * machine_code_context = nullptr;
+
+        if (target_machine.addPassesToEmitMC(pass_manager, machine_code_context, object_stream))
+            throw Exception(ErrorCodes::CANNOT_COMPILE_CODE, "MachineCode is not supported for the platform");
+
+        pass_manager.run(module);
+
+        std::unique_ptr<llvm::MemoryBuffer> compiled_object_buffer = std::make_unique<llvm::SmallVectorMemoryBuffer>(
+            std::move(object_buffer), "<in memory object compiled from " + module.getModuleIdentifier() + ">");
+
+        return compiled_object_buffer;
+    }
+
+    ~JITCompiler() = default;
+
+private:
+    llvm::TargetMachine & target_machine;
+};
+
+/** MemoryManager for module.
+  * Keep total allocated size during RuntimeDyld linker execution.
+  * Release compiled module memory in destructor.
```

Comment:
Don't see the destructor...

Reply:
@alexey-milovidov it is done by parent class, in comments this is mentioned because we put module manager in hash map to keep generated code valid until it will be removed.